Remove unsupported yarn install option. Default to use Yarn berry

saurori opened the pull request:
Resolves this Issue: https://github.com/gobuffalo/cli/issues/95

Yarn 2+ (berry) is not supported on `buffalo new`. These changes allow for Yarn 1.x (classic) and 2.x+ (berry) to work, and also defaults to using the latest Yarn berry on setup.

sio4 commented on internal/cmd/setup/frontend.go:
```diff
@@ -12,12 +12,12 @@ func yarnCheck(app meta.App) error {
 	}
 	if _, err := exec.LookPath("yarnpkg"); err != nil {
-		err := run(exec.Command("npm", "install", "-g", "yarn"))
+		err := run(exec.Command("npm", "install", "-g", "yarn@berry"))
```
This code is for the `buffalo setup` command which is used not only for a newly created app but also for an existing app such as just checked out from its repository to a new dev environment. In the latter case, we cannot guarantee the app is configured for berry and I guess this approach is not the best solution. I think we can keep the original line here. Do you check such a scenario?

saurori replied:
That sounds good, will revert.

sio4 commented on internal/cmd/setup/frontend.go:
```diff
@@ -12,12 +12,12 @@ func yarnCheck(app meta.App) error {
 		}
 	}
-	if err := run(exec.Command("yarnpkg", "install", "--no-progress")); err != nil {
+	if err := run(exec.Command("yarnpkg", "install", "--silent")); err != nil {
```
`--silent` option suppresses the whole output for classic yarn, and this is not good. With the results of some searching and testing, I think we can just remove these options.

sio4 commented on internal/genny/assets/webpack/webpack.go:
```diff
@@ -88,15 +88,19 @@ func New(opts *Options) (*genny.Generator, error) {
 
 func installPkgs(r *genny.Runner, opts *Options) error {
 	command := "yarnpkg"
+	args := []string{"install", "--silent"}
```
Same for the above comment (`--silent`)

sio4 commented on internal/genny/assets/webpack/webpack.go:
```diff
@@ -88,15 +88,19 @@ func New(opts *Options) (*genny.Generator, error) {
 	}
-	args := []string{"install", "--no-progress", "--save"}
 
 	c := exec.Command(command, args...)
```
Since the command generates more than 4K lines of logs for yarn, which is too much verbose, we should add `logFilters` to `.yarnrc.yml` that I mentioned in the issue comment before executing the command. I tried to find the equivalent command-line options or environment variables but could not find them.

sio4 commented on internal/genny/docker/templates/dot-dockerignore.tmpl:
```diff
@@ -1,3 +1,4 @@
 node_modules/
+.pnp.*
```
`.yarn` too maybe?

saurori replied:
Yeah I guess everything that is in the gitignore template should also be in dockerignore, per: https://yarnpkg.com/getting-started/qa#which-files-should-be-gitignored

saurori replied:
I don't think it really matters. The [Dockerfile template](https://github.com/gobuffalo/cli/blob/v0.18.2/internal/genny/docker/templates/Dockerfile.tmpl) uses a multi-stage dockferfile where the assets and binary are built in a `builder` image and then copied over to the run image. The yarn files will only be present in the `builder` image once complete.

sio4 replied:
Ah! right! I missed that. You are correct.

sio4 reviewed (changes requested):
Please check some comments on the code and please take a look at the comment on the original issue #95

saurori commented:
@sio4 I pushed fixes addressing your comments https://github.com/gobuffalo/cli/pull/96/commits/62585b46f2215ddaa8376c7867307d7446b9551c. I also added `enableGlobalCache: true` to the `.yarnrc.yml` file as I think it's a good default to avoid duplicating dependencies across projects.

The resulting `.yarnrc.yml` file after running `buffalo new` looks like:
```yaml
enableGlobalCache: true

logFilters:
  - code: YN0013
    level: discard

yarnPath: .yarn/releases/yarn-3.2.0.cjs
```

sio4 commented:
Thank you for your update @saurori ! I think we need to check the .dockerignore file part, meanwhile, I will check the updated branch and will add a comment soon (maybe tomorrow)!

saurori commented:
@sio4 any other changes? Are you able to approve the workflow (tests)?

saurori commented:
I'm not sure what is causing tests failures:
```console
=== RUN   Test_InstallOldBuffaloCMD/existing_version
    testhelpers_test.go:63: 
        	Error Trace:	testhelpers_test.go:63
        	Error:      	Received unexpected error:
        	            	unknown gobuffalo cli version v0.18.1
        	Test:       	Test_InstallOldBuffaloCMD/existing_version
=== RUN   Test_InstallOldBuffaloCMD/existing_old_version
    testhelpers_test.go:63: 
        	Error Trace:	testhelpers_test.go:63
        	Error:      	Received unexpected error:
        	            	unknown gobuffalo cli version v0.16.27
        	Test:       	Test_InstallOldBuffaloCMD/existing_old_version
```

sio4 commented:
Sorry for the delay. I was somewhat busy. It seems like testing was done now. I will check if there is something more and will also check the test failure.

sio4 reviewed (approved):
LGTM

sio4 commented:
I found the reason for the failure by PRing #110. The issue was, I guess, the underlying API call for the test case has a rate limit and we sent too many API calls to api.github.com. Working on it.

sio4 commented:
This could be tested again once #110 is merged into `development`.

paganotoni commented:
@saurori can you please rebase your branch with development? Thanks in advance

saurori commented:
@paganotoni rebased

saurori commented:
I am not 100% sure why this failed:

```console
=== RUN   TestBuild/skipop
...
DEBU[2022-03-10T21:34:17Z] Exec: yarnpkg install
        DEBU[2022-03-10T21:34:17Z] ➤ YN0028: The lockfile would have been created by this install, which is explicitly forbidden.
DEBU[2022-03-10T21:34:17Z] ➤ YN0000: Failed with errors in 0s 10ms
```

Running integration tests locally passes. The error doesn't make sense for a first time `yarn install` unless immutable builds is set to true. The install command isn't using the `--immutable` flag, so the only other way that setting is enabled is via env var `YARN_ENABLE_IMMUTABLE_INSTALLS=true` which means it won't install without a `yarn.lock` file.

`DEBU[2022-03-10T21:34:14Z] LookPath: yarnpkg` indicates that yarn is present and does not need to be installed. I wonder if the version of yarn used is the problem? Both the [mac](https://github.com/actions/virtual-environments/blob/main/images/macos/macos-11-Readme.md) and [ubuntu](https://github.com/actions/virtual-environments/blob/releases/ubuntu20/20220306/images/linux/Ubuntu2004-Readme.md) github virtual environments list `Yarn 1.22.17`

My local Yarn:
```console
$ yarnpkg --version
2.4.3
```

sio4 commented:
Checking...

sio4 commented:
> The error doesn't make sense for a first time `yarn install` unless immutable builds is set to true. The install command isn't using the `--immutable` flag, so the only other way that setting is enabled is via env var `YARN_ENABLE_IMMUTABLE_INSTALLS=true` which means it won't install without a `yarn.lock` file.

Oh, I found the direct cause: As you say, this symptom is about immutable and I found that:

> If the `--immutable` option is set (**defaults to true on CI**), Yarn will abort with an error exit code if the lockfile was to be modified (other paths can be added using the `immutablePatterns` configuration setting). For backward compatibility we offer an alias under the name of `--frozen-lockfile`, but it will be removed in a later release.

This behavior of berry is annoying. It seems like there is no opposite option of `--immutable` such as `--no-immutable`. (Could be a good FR point?) but the only way I found is using configuration.

Here is my quick patch which covers immutable issue, yarnrc.yml version issue when system yarn is still classic, and consolidate them in a single function. Could you please take a look at this and try applying it if you want? Or I will PR to your branch later. (I have no enough time today)

```diff
diff --git a/internal/genny/assets/webpack/templates/dot-yarnrc.yml.tmpl b/internal/genny/assets/webpack/templates/dot-yarnrc.yml.tmpl
deleted file mode 100644
index ba06a62..0000000
--- a/internal/genny/assets/webpack/templates/dot-yarnrc.yml.tmpl
+++ /dev/null
@@ -1,4 +0,0 @@
-enableGlobalCache: true
-logFilters:
-  - code: YN0013
-    level: discard
diff --git a/internal/genny/assets/webpack/webpack.go b/internal/genny/assets/webpack/webpack.go
index 0c0fe5c..68fbb55 100644
--- a/internal/genny/assets/webpack/webpack.go
+++ b/internal/genny/assets/webpack/webpack.go
@@ -13,6 +13,8 @@ import (
 	"github.com/gobuffalo/genny/v2/gogen"
 )
 
+const YARN = "yarnpkg"
+
 //go:embed templates/* templates/assets/css/_buffalo.scss.tmpl
 var templates embed.FS
 
@@ -41,7 +43,7 @@ func New(opts *Options) (*genny.Generator, error) {
 
 	g.RunFn(func(r *genny.Runner) error {
 		if opts.App.WithYarn {
-			if _, err := r.LookPath("yarnpkg"); err == nil {
+			if _, err := r.LookPath(YARN); err == nil {
 				return nil
 			}
 			// If yarn is not installed, it still can be installed with npm.
@@ -87,7 +89,7 @@ func New(opts *Options) (*genny.Generator, error) {
 }
 
 func installPkgs(r *genny.Runner, opts *Options) error {
-	command := "yarnpkg"
+	command := YARN
 	args := []string{"install"}
 
 	if !opts.App.WithYarn {
@@ -97,11 +99,12 @@ func installPkgs(r *genny.Runner, opts *Options) error {
 		if err := installYarn(r); err != nil {
 			return err
 		}
-		if err := r.Exec(exec.Command(command, []string{"set", "version", "berry"}...)); err != nil {
+		if err := setupYarnBerry(r); err != nil {
 			return err
 		}
 	}
 
+	r.Exec(exec.Command(command, "--version"))
 	c := exec.Command(command, args...)
 	c.Stdout = yarnWriter{
 		fn: r.Logger.Debug,
@@ -123,9 +126,31 @@ func (y yarnWriter) Write(p []byte) (int, error) {
 
 func installYarn(r *genny.Runner) error {
 	// if there's no yarn, install it!
-	if _, err := r.LookPath("yarnpkg"); err == nil {
+	if _, err := r.LookPath(YARN); err == nil {
 		return nil
 	}
 	yargs := []string{"install", "-g", "yarn"}
 	return r.Exec(exec.Command("npm", yargs...))
 }
+
+func setupYarnBerry(r *genny.Runner) error {
+	if err := r.Exec(exec.Command(YARN, "--version")); err != nil {
+		return err
+	}
+	if err := r.Exec(exec.Command(YARN, "set", "version", "berry")); err != nil {
+		return err
+	}
+	// the .yarnrc.yml format is different between classic and berry so
+	// it should be configured after installation (set version)
+	if err := r.Exec(exec.Command(YARN, "config", "set", "enableGlobalCache", "true")); err != nil {
+		return err
+	}
+	// can we enable this only for testing? :-(
+	if err := r.Exec(exec.Command(YARN, "config", "set", "enableImmutableInstalls", "true")); err != nil {
+		return err
+	}
+	if err := r.Exec(exec.Command(YARN, "config", "set", "logFilters", "--json", `[{"code":"YN0013","level":"discard"}]`)); err != nil {
+		return err
+	}
+	return nil
+}
```

sio4 commented:
I made a PR to your branch (saurori/cli/pull/1) Please take a look at the PR.

saurori commented:
@paganotoni if you could re-run the CI with the latest change, thanks!

sio4 commented on internal/genny/assets/webpack/templates/dot-yarnrc.yml.tmpl:
```diff
@@ -0,0 +1,4 @@
+enableGlobalCache: true
+logFilters:
```
Could you please consider using `yarn config set` instead of this template? It doesn't help reducing 4k lines of `yarn install` logs if home version is not berry.

saurori replied:
Ok I see now. Yarn classic will only look for `.yarnrc` not `.yarnrc.yml`, so using `yarn config set` will work on both versions.

saurori replied:
Updated in https://github.com/gobuffalo/cli/pull/96/commits/4858a22a319fa7d641d87356c4112c9ae8c60488. 🤞 that solves everything

sio4 reviewed (changes requested):
Please consider the situation when if user's version is classic.

paganotoni commented:
Seems is passing the tests now 🥳  :)

sio4 commented:
Yeah, to be clarified, the routine added now into cli is only for the new app with berry, so this is not about compatibility with the classic version. The `yarn config set ...` command will always be run with berry but `yarn set version ...` could be invoked with the classic version if the user's global `yarn` is classic. 

The classic version had no `version set version` before. It was added to support berry installation from classic [1] and the command will overwrite (existing) `.yarnrc.yml` [2] so the template will be ruined. This is the reason I asked to use `yarn config` **after** the new `.yarnrc.yml` created by both versions of yarn.

[1] https://github.com/yarnpkg/yarn/commit/f381c5f0ce85a96dc1cba8e4883819ee878c949f
[2] https://github.com/yarnpkg/yarn/blob/6db39cf0ff684ce4e7de29669046afb8103fce3d/src/cli/commands/policies.js#L166:L170

sio4 reviewed (approved):
lgtm!
